File: PROYECTO ARBOL AVL/Estacionamiento.cpp
#include "Estacionamiento.h"

Estacionamiento::Estacionamiento() {
    std::srand(0);
}

void Estacionamiento::ocuparEspacio(int espacio, Coche& coche) {
    if (espacio >= 0 && espacio < TAMANIO && !espacioOcupado(espacio)) {
        espaciosOcupados[espacio] = coche;  
       

    } else {
        std::cout << "Espacio no válido o ya ocupado." << std::endl;
    }
}

void Estacionamiento::liberarEspacio(const std::string& placa) {
    try {
        
        int espacio = obtenerEspacioPorPlaca(placa); 
        
       
        if (espacioOcupado(espacio)) {
            espaciosOcupados.erase(espacio);  
            std::cout << "Espacio " << espacio << " liberado." << std::endl;
        } else {
            std::cout << "Espacio no ocupado." << std::endl;
        }
    } catch (const std::invalid_argument& e) {
        std::cout << e.what() << std::endl; 
    }
}

int Estacionamiento::obtenerEspacioPorPlaca(const std::string& placa) {
   
   std::cout << "=== DEBUG: Contenido de espaciosOcupados ===" << std::endl;
if (espaciosOcupados.empty()) {
    std::cout << "El mapa está vacío, no hay coches almacenados." << std::endl;
} else {
    for (const auto& pair : espaciosOcupados) {
        std::cout << "Espacio: " << pair.first << " | Placa: " << pair.second.getPlaca() << std::endl;
    }
}
std::cout << "==========================================" << std::endl;
    for (const auto& pair : espaciosOcupados) {
        const Coche& coche = pair.second;
        std::cout << "  Comparando con placa: " << coche.getPlaca() 
                  << " en espacio " << pair.first << std::endl;

        if (coche.getPlaca() == placa) {
            std::cout << "  -> Coche encontrado en el espacio " << pair.first << std::endl;
            return pair.first;
        }
    }

    std::cout << "  -> Coche NO encontrado." << std::endl;
    throw std::invalid_argument("Coche no encontrado en el estacionamiento.");
}



void Estacionamiento::mostrarEstacionamiento() {
    std::cout << "Estado del Parqueadero:\n";
    
    
    std::cout << "Entrada            | Salida 1 | Salida 2\n";
    std::cout << "------------------------------------------\n";
    
   
    for (int i = 0; i < TAMANIO; ++i) {
       
        if (espacioOcupado(i)) {
            std::cout << "[X] ";  
        } else {
            std::cout << "[ ] "; 
        }

        if ((i + 1) % 5 == 0) {
            std::cout << "   ";  
        }

      
        if ((i + 1) == 5) {
            std::cout << "| Salida 1 ";
        } else if ((i + 1) == 95) {
            std::cout << "| Salida 2 ";
        }

       
        if ((i + 1) % 10 == 0) {
            std::cout << std::endl;
        }
    }

    std::cout << std::endl;
}

int Estacionamiento::obtenerEspacioAleatorio() {
    int espacio = rand() % TAMANIO;  
    
    while (espacioOcupado(espacio)) {  
        espacio = rand() % TAMANIO;  
    }
    
    return espacio;  
}

int Estacionamiento::getNumeroEspacios() const {
    return TAMANIO; 
}

bool Estacionamiento::espacioOcupado(int espacio) {
    return espaciosOcupados.find(espacio) != espaciosOcupados.end();  
}

Coche Estacionamiento::obtenerCocheEnEspacio(int espacio) {
    if (espacioOcupado(espacio)) {
        return espaciosOcupados[espacio];  
    } else {
        throw std::invalid_argument("El espacio no está ocupado.");
    }
}

void Estacionamiento::vaciarEstacionamiento() {
    espaciosOcupados.clear();  
    std::cout << "Estacionamiento vaciado correctamente." << std::endl;
}

File: PROYECTO PYTHON/Estacionamiento.py
import random


class Coche:
    def __init__(self, placa):
        self.placa = placa
    
    def get_placa(self):
        return self.placa


class Estacionamiento:
    TAMANIO = 100  # Tamaño del estacionamiento

    def __init__(self):
        self.espacios_ocupados = {}

    def ocupar_espacio(self, espacio, coche):
        if 0 <= espacio < self.TAMANIO and not self.espacio_ocupado(espacio):
            self.espacios_ocupados[espacio] = coche
        else:
            print("Espacio no válido o ya ocupado.")

    def liberar_espacio(self, placa):
        try:
            espacio = self.obtener_espacio_por_placa(placa)
            if self.espacio_ocupado(espacio):
                del self.espacios_ocupados[espacio]
                print(f"Espacio {espacio} liberado.")
            else:
                print("Espacio no ocupado.")
        except ValueError as e:
            print(e)

    def obtener_espacio_por_placa(self, placa):
        print("=== DEBUG: Contenido de espaciosOcupados ===")
        if not self.espacios_ocupados:
            print("El mapa está vacío, no hay coches almacenados.")
        else:
            for espacio, coche in self.espacios_ocupados.items():
                print(f"Espacio: {espacio} | Placa: {coche.get_placa()}")
        print("==========================================")

        for espacio, coche in self.espacios_ocupados.items():
            print(f"  Comparando con placa: {coche.get_placa()} en espacio {espacio}")
            if coche.get_placa() == placa:
                print(f"  -> Coche encontrado en el espacio {espacio}")
                return espacio

        print("  -> Coche NO encontrado.")
        raise ValueError("Coche no encontrado en el estacionamiento.")

    def mostrar_estacionamiento(self):
        print("Estado del Parqueadero:")
        for i in range(self.TAMANIO):
            if self.espacio_ocupado(i):
                print("[X]", end=" ")  # Espacio ocupado
            else:
                print("[ ]", end=" ")  # Espacio libre

            if (i + 1) % 5 == 0:
                print("   ", end="")  # Espacios visuales

            if (i + 1) % 10 == 0:
                print()  # Nueva línea cada 10 espacios

        print()

    def obtener_espacio_aleatorio(self):
        espacio = random.randint(0, self.TAMANIO - 1)

        while self.espacio_ocupado(espacio):
            espacio = random.randint(0, self.TAMANIO - 1)

        return espacio

    def get_numero_espacios(self):
        return self.TAMANIO

    def espacio_ocupado(self, espacio):
        return espacio in self.espacios_ocupados

    def obtener_coche_en_espacio(self, espacio):
        if self.espacio_ocupado(espacio):
            return self.espacios_ocupados[espacio]
        else:
            raise ValueError("El espacio no está ocupado.")

    def vaciar_estacionamiento(self):
        self.espacios_ocupados.clear()
        print("Estacionamiento vaciado correctamente.")
